File: 백준/Gold/3109. 빵집/빵집.py
R, C = map(int, input().split())
board = []
for _ in range(R):
    b = list(input())
    board.append(b)

def dfs(i, j):
    global board
    if i >= len(board) or i < 0:
        return False
    if j >= len(board[0]) or j < 0:
        return False
    if board[i][j] == "x":
        return False
    if j == len(board[0])-1:
        return True

    Di = [-1, 0, 1]
    for di in Di:
        board[i][j] = "x"
        if dfs(i+di, j+1):
            return True
    return False

count = 0
for i in range(R):
    if dfs(i, 0):
        count += 1

print(count)

File: 백준/Gold/3109. 빵집/빵집.java
import java.io.*;
import java.util.*;

public class Main {

    static int R, C;
    static char[][] board;
    static int ans;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        R = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());

        board = new char[R][C];
        for(int i=0;i<R;i++) {
            String[] row = br.readLine().split("");
            for(int j=0;j<C;j++) {
                board[i][j] = row[j].charAt(0);
            }
        }

        ans = 0;
        for(int i=0;i<R;i++) {
            dfs(i, 0);
        }

        System.out.println(ans);
    }

    private static boolean dfs(int y, int x) {
        if(x == C-1) {
            ans++;
//            for(int i=0;i<R;i++)
//                System.out.println(Arrays.toString(board[i]));
            return true;
        }

        if(y-1 >= 0 && board[y-1][x+1] == '.') {
            board[y-1][x+1] = 'O';
            if(dfs(y-1, x+1)) {
                return true;
            }
            board[y - 1][x + 1] = '.';
        }
        if(board[y][x+1] == '.') {
            board[y][x+1] = 'O';
            if(dfs(y, x+1)) {
                return true;
            }
            board[y][x + 1] = '.';
        }
        if(y+1 < R && board[y+1][x+1] == '.') {
            board[y+1][x+1] = 'O';
            if(dfs(y+1, x+1)) {
                return true;
            }
        }

        return false;
    }
}